File: python/234. Palindrome Linked List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        current = head
        middleCurrent = self.reverseLinkedList(slow)

        while middleCurrent:
            if current.val != middleCurrent.val:
                return False
            current = current.next
            middleCurrent = middleCurrent.next


        return True


    def reverseLinkedList(self, head):

        prev = None
        current = head

        while current:
            temp = current.next
            current.next = prev
            prev = current
            current = temp

        return prev

File: java/234. Palindrome Linked List/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null) return true;

        ListNode current = head;
        ListNode middle = head;

        while (current != null && current.next != null) {
            middle = middle.next;
            current = current.next.next;
        }

        current = head;
        ListNode midCurrent = reverseLinkedList(middle);

        while(midCurrent != null) {
            if (current.val != midCurrent.val) return false;
            current = current.next;
            midCurrent = midCurrent.next;
        }
        return true;
    }

    public ListNode reverseLinkedList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;

        while (current != null) {
            ListNode tmp = current.next;
            current.next = prev;
            prev = current;
            current = tmp;
        }

        return prev;

    }
}